File: RAPLRust/warmup/src/Rust/rapl-interface/benchmarks/fibsequence/java/Bench.java
import java.lang.foreign.*;
import java.lang.invoke.MethodHandle;

class Bench {
    public static void main(String[] args) {

        // Finding os
        var os = System.getProperty("os.name");

        // Finding the path of library (and loading it)
        var dll_path = System.getProperty("user.dir") + "/target/release/";
        if (os.equals("Linux")) {
            dll_path = dll_path + "librapl_lib.so";
        } else if (os.equals("Windows 11")) {
            dll_path = dll_path + "rapl_lib.dll";
        } else {
            System.out.println("OS not supported");
            return;
        }

        System.load(dll_path);

        // Loading functions
        MemorySegment start_rapl_symbol = SymbolLookup.loaderLookup().find("start_rapl").get();
        MethodHandle start_rapl = Linker.nativeLinker().downcallHandle(start_rapl_symbol,
                    FunctionDescriptor.of(ValueLayout.JAVA_INT));

        MemorySegment stop_rapl_symbol = SymbolLookup.loaderLookup().find("stop_rapl").get();
        MethodHandle stop_rapl = Linker.nativeLinker().downcallHandle(stop_rapl_symbol,
                    FunctionDescriptor.of(ValueLayout.JAVA_INT));

        
        // Getting arguments
        int loop_count = Integer.parseInt(args[0]);
        int n = Integer.parseInt(args[1]);

        // Running benchmark
        // Note that this could potentially be optimized away
        // by the compiler due to the fact that the result is not used.
        for (int i = 0; i < loop_count; i++) {
            try {
                start_rapl.invoke();
            } catch (Throwable e) {
                e.printStackTrace();
            }

            long result = recFibN(n);

            try {
                stop_rapl.invoke();
            } catch (Throwable e) {
                e.printStackTrace();
            }
            
            // stopping compiler optimization
            if (result < 42){
                System.out.println(result);
            }
        }
    }

    // test method from Rosetta code
    public static long recFibN(final int n)
    {
        return (n < 2) ? n : recFibN(n - 1) + recFibN(n - 2);
    }
}

File: RAPLRust/lib/rapl-interface/benchmarks/fibsequence/javascript/bench.js
const os = require("os");

// getting arguments
const runCount = process.argv[2];
const fibParam = process.argv[3];

// finding path depending on OS
const libPath = os.platform() == "win32" ?
  "target\\release\\rapl_lib.dll" :
  "target/release/librapl_lib.so"

// test method from Rosetta code
function fib(n) {
  return n<2?n:fib(n-1)+fib(n-2);
}

// loading library
const koffi = require('koffi');
const lib = koffi.load(libPath);

// loading functions
const start = lib.func('int start_rapl()');
const stop = lib.func('void stop_rapl()');

// running benchmark
for (let i = 0; i < runCount; i++) {
  start();

  let result = fib(fibParam);

  stop();
  if (result < 42){
      console.log(result);
  }
}
